### PR TITLE
(feat): wilcoxon test

### PR DESCRIPTION
**Pull request by ilan-gold**
I had to use Mann Whitney because there's no guarantee on size of the inputs, right?  This is what `diffxpy` does.

A few things I noticed:
1. It would be nice to test correctness - how do we know which columns of `test_adata` have different distributions or not?
2. Not only does `fit` not do anything, but I'm not even sure what `design` is for in the constructor.  Not opposed to leaving it, just noting.
3. Does `check_counts` actually need to do anything here?  The test doesn't actually care about the input, although maybe it would be good to check for consistency.  But maybe people use this test on z-scored things (I don't know).
4. Should we change the name of the test?
5. Should we be testing different data types (sparse vs. not)?

### CODE REVIEW

File: src/multi_condition_comparisions/tl/de.py [hunk 1]
```diff
@@ -466,3 +467,39 @@ def _test_single_contrast(self, contrast: list[str], **kwargs) -> pd.DataFrame:
         de_res = ro.conversion.rpy2py(ro.globalenv["de_res"])
 
         return de_res
+    
+
+class WilcoxonTest(BaseMethod):
+    def _check_counts(self) -> bool:
+        return True # later? is this correct?
```
Comment (Zethson) [inline src/multi_condition_comparisions/tl/de.py:474]: I think that people are also using the scanpy implementation very flexibly. So yaaa probably fine?

File: src/multi_condition_comparisions/tl/de.py [hunk 2]
```diff
@@ -466,3 +467,39 @@ def _test_single_contrast(self, contrast: list[str], **kwargs) -> pd.DataFrame:
...
+    
+    def fit(self):
+        warnings.warn("There is nothing to fit in a wilcoxon test.")
```
Comment (Zethson) [inline src/multi_condition_comparisions/tl/de.py:477]: Should this raise a NotImplementedError? Or a ValueError?

File: src/multi_condition_comparisions/tl/de.py [hunk 3]
```diff
@@ -466,3 +467,39 @@ def _test_single_contrast(self, contrast: list[str], **kwargs) -> pd.DataFrame:
...
+            adata0 = self.adata[self.adata.obs[contrast[0]] == contrast[1], var]
+            adata1 = self.adata[self.adata.obs[contrast[0]] == contrast[2], var]
+
+            x0 = adata0.X if self.layer is None else adata0.layers[self.layer]
+            x1 = adata1.X if self.layer is None else adata1.layers[self.layer]
+            pval = scipy.stats.mannwhitneyu(
+                x=np.asarray(x0.todense()).flatten() if scipy.sparse.issparse(x0) else x0,
+                y=np.asarray(x1.todense()).flatten() if scipy.sparse.issparse(x1) else x1,
+                use_continuity=True,
+                alternative="two-sided"
```
Comment (Zethson) [inline src/multi_condition_comparisions/tl/de.py:494]: Should this be a kwarg?

**Review by Zethson — COMMENTED** [review]
> It would be nice to test correctness - how do we know which columns of test_adata have different distributions or not?

I guess we just have to create an artificial AnnData object where a few genes differ so strongly that they have to be DE. Like over the top.

> Not only does fit not do anything, but I'm not even sure what design is for in the constructor. Not opposed to leaving it, just noting.

Nothing I guess.

Does check_counts actually need to do anything here? The test doesn't actually care about the input, although maybe it would be good to check for consistency. But maybe people use this test on z-scored things (I don't know).

> Pretty sure that the Wilcoxon test is used for absolutely everything given how people use it in their own packages (people that build on top of scanpy's rank genes).

> Should we change the name of the test?

To?

> Should we be testing different data types (sparse vs. not)?

Yeah I would.

**Comment by grst** [comment]
> Not only does fit not do anything, but I'm not even sure what design is for in the constructor. Not opposed to leaving it, just noting.

Maybe you are right, and I'm sacrifying user experience for consistency. But with the exception of benchmarking there's no good reason to have it consistent at any cost?

So maybe we should go for a simpler interface in the end, something like

```python
test_wilcoxon(adata, group_col: str, paired_col: str=None, comparisons={"a_vs_b": ["A", "B"], "a_vs_rest": ["A", "rest"]})
```

I'm really unsure what's better. @Zethson, WDYT?

> I had to use Mann Whitney because there's no guarantee on size of the inputs, right?

I'm always confused by the naming... Mann-Whitney is for two independent groups. The wilcoxon test is for paired observations. We should have both (same as for t-test).

File: src/multi_condition_comparisions/tl/de.py (hunk 2 above)
Comment (grst) [reply src/multi_condition_comparisions/tl/de.py:477]: Nope, if we are to keep this interface, this should just be a `pass`. 

Consider something like: 

```python
methods = {"wilcox": WilcoxonTest, "lm": StatsmodelsDE}
results = {}
for method_name, method in methods.items():
    model = method("~variable")
    model.fit()
    results[method_name] = model.test_contrasts({"a_vs_b": model.contrast("variable", "A", "B")})
```

File: src/multi_condition_comparisions/tl/de.py [hunk 4]
```diff
@@ -466,3 +467,39 @@ def _test_single_contrast(self, contrast: list[str], **kwargs) -> pd.DataFrame:
...
+
+class WilcoxonTest(BaseMethod):
+    def _check_counts(self) -> bool:
+        return True # later? is this correct?
+    
+    def fit(self):
+        warnings.warn("There is nothing to fit in a wilcoxon test.")
+
+
+    def _test_single_contrast(self, contrast, **kwargs) -> pd.DataFrame:
```
Comment (grst) [inline src/multi_condition_comparisions/tl/de.py:480]: If this is supposed to be compatible with the other classes, `contrast` will be a contrast vector matching a design matrix. 
I'm sure this could be made to work, but as noted in my other comment, it might not be a good idea after all.

**Comment by grst** [comment]
>  how do we know which columns of test_adata have different distributions or not?

@BorisMuzellec, is there any description about the pydeseq2 test data? 
https://github.com/scverse/multi-condition-comparisions/blob/6907fb665054dadaab5a1d9db952f2b7b08211f5/tests/conftest.py#L10-L14

**Comment by BorisMuzellec** [comment]
> @BorisMuzellec, is there any description about the pydeseq2 test data?

Hi @grst, here's the R code I used to sample the data (it's a 2-factor modification of DESeq2's `makeExampleDESeqDataSet`)

```
library(DESeq2)

makeExampleDESeqDataSet <-
  function(n = 10,
           m = 100,
           betaSD = 0,
           interceptMean = 4,
           interceptSD = 2,
           dispMeanRel = function(x)
             4 / x + .1,
           sizeFactors = rep(1, m)) {
    beta <-
      cbind(rnorm(n, interceptMean, interceptSD),
            rnorm(n, 0, betaSD),
            rnorm(n, 0, betaSD))
    dispersion <- dispMeanRel(2 ^ (beta[, 1]))
    colData <-
      DataFrame(condition = factor(rep(c("A", "B"), times = c(
        ceiling(m / 2), floor(m / 2)
      ))))
    colData$group <- factor(rep(c("X", "Y"), times = m / 2))
    
    x <- if (m > 1) {
      stats::model.matrix.default( ~ colData$group + colData$condition)
    } else {
      cbind(rep(1, m), rep(0, m))
    }
    mu <- t(2 ^ (x %*% t(beta)) * sizeFactors)
    countData <-
      matrix(rnbinom(m * n, mu = mu, size = 1 / dispersion), ncol = m)
    mode(countData) <- "integer"
    colnames(countData) <- paste("sample", 1:m, sep = "")
    rowRanges <-
      GRanges("1", IRanges(start = (1:n - 1) * 100 + 1, width = 100))
    names(rowRanges) <- paste0("gene", 1:n)
    
    # set environment to global environment,
    # to avoid the formula carrying with it all the objects
    # here including 'object' itself.
    design <- if (m > 1) {
      as.formula("~ group + condition", env = .GlobalEnv)
    } else {
      as.formula("~ 1", env = .GlobalEnv)
    }
    
    object <- DESeqDataSetFromMatrix(
      countData = countData,
      colData = colData,
      design = design,
      rowRanges = rowRanges
    )
    trueVals <- DataFrame(trueIntercept = beta[, 1],
                          trueBeta = beta[, 2],
                          trueDisp = dispersion)
    mcols(trueVals) <- DataFrame(
      type = rep("input", ncol(trueVals)),
      description = c(
        "simulated intercept values",
        "simulated beta values",
        "simulated dispersion values"
      )
    )
    mcols(object) <- cbind(mcols(object), trueVals)
    return(object)
  }

# Sample a dummy DEA dataset
set.seed(1)
dds <- makeExampleDESeqDataSet(m = 100, n = 10, betaSD = .25)
```

File: src/multi_condition_comparisions/tl/de.py (hunk 4 above)
Comment (ilan-gold) [reply src/multi_condition_comparisions/tl/de.py:490]: What is the context of this comment @grst ? Not totally clear.

Comment (grst) [reply src/multi_condition_comparisions/tl/de.py:490]: It relates to the general question if we should force-fit wilcoxon test into the same API as linear models for the sake of consistency or rather use a different API that is better suited for simple tests. See https://github.com/scverse/multi-condition-comparisions/pull/31#issuecomment-1874964212 / https://github.com/scverse/multi-condition-comparisions/issues/25#issuecomment-1871447103 for arguments. 

I don't think we made a decision here yet?

Comment (ilan-gold) [reply src/multi_condition_comparisions/tl/de.py:490]: I thought we had.   I missed the bit about erroring for the design but other than that, seems like we landed on something.  I am a fan of keeping the API uniform even if it semantically doesn't necessarily make sense.  I guess another option for API uniformity would be to cache the result of `fit` after running it on the first `test_contrasts`.  Maybe I'm missing some update that is possible to `model` though that would necessitate a refitting?

Comment (grst) [reply src/multi_condition_comparisions/tl/de.py:490]: ok, in that case the point is that for the linear models, the contrast passed to this function will always be a vector such as 

```python
contrast = [0, 0, 1]
# or
contrast = [0, -1, 1]
# or 
contrast = [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, -1]
```

where the length of `contrast` is equivalent to the number of columns in the design matrix. 

The base class provides tooling to build such contrast vectors, e.g. 

```python
# simple contrast
contrast = model.contrast(column="group", baseline="A", group_to_compare="B")
# for more complex contrasts
contrast = model.cond(cell = "B cells", condition = "stim") - model.cond(cell = "B cells", condition = "ctrl")
```

File: src/multi_condition_comparisions/tl/de.py [hunk 5]
```diff
@@ -466,3 +468,65 @@ def _test_single_contrast(self, contrast: list[str], **kwargs) -> pd.DataFrame:
...
+    def fit(self):
+        """`fit` method here is only provided for API completeness and does nothing."""
+        warnings.warn(f"There is nothing to fit in a {self.name} test.", stacklevel=2)
+        pass
+
+    def _test_single_contrast(self, contrast, **kwargs) -> pd.DataFrame:
+        res = []
+        if len(contrast) != 3:
+            raise ValueError("Contrast can only have three elements for Mann-Whitney U test.")
+        for var in tqdm(self.adata.var_names):
```
Comment (grst) [inline src/multi_condition_comparisions/tl/de.py:494]: consider using `tqdm.contrib.concurrent.process_map` instead of `tqdm` for "free" parallelization

File: src/multi_condition_comparisions/tl/de.py [hunk 6]
```diff
@@ -466,3 +468,65 @@ def _test_single_contrast(self, contrast: list[str], **kwargs) -> pd.DataFrame:
         de_res = ro.conversion.rpy2py(ro.globalenv["de_res"])
 
         return de_res
+
+
+class BaseTwoSampleNonParametricTest(BaseMethod):
```
[the remark this answers is not in the document]
Comment (ilan-gold) [reply src/multi_condition_comparisions/tl/de.py:473]: Maybe I misunderstand, but isn't that the difference between wilcoxon and mann-whitney?

[the remark this answers is not in the document]
Comment (grst) [reply src/multi_condition_comparisions/tl/de.py:473]: yeah, but they could have the same interface? But even if they are separate classes, you still need a way to specify the grouping variable for the mann-whitney test.

[the remark this answers is not in the document]
Comment (grst) [reply src/multi_condition_comparisions/tl/de.py:473]: not sure tbh... paired design is just one special case of a formula. Maybe we could also infer from the design matrix and check if it's a paired design, and error if it's not.

[the remark this answers is not in the document]
Comment (ilan-gold) [reply src/multi_condition_comparisions/tl/de.py:473]: I think that gets back to the same thing though, no?  How would you distinguish between `n/2` classes and pairing?  Or is that actually the same thing?  I think the test depends on how you would interpret the class labels.

[the remark this answers is not in the document]
Comment (grst) [reply src/multi_condition_comparisions/tl/de.py:473]: in the context of a linear model, pairing is the same as n/2 classes with 2 samples each.

[the remark this answers is not in the document]
Comment (ilan-gold) [reply src/multi_condition_comparisions/tl/de.py:473]: @grst I think I need some clarification here.  These tests are basically meant for two groups.  In the context `~ group + paired_by`, assuming `group` and `paired_by` have multiple subgroups, at least 4 combinations.  How can we infer based on these if something is paired or not?  I think this could only be done at `test_contrasts` no?

[the remark this answers is not in the document]
Comment (grst) [reply src/multi_condition_comparisions/tl/de.py:473]: agree this probably needs to be done at the `test_contrasts` stage.

[the remark this answers is not in the document]
Comment (ilan-gold) [reply src/multi_condition_comparisions/tl/de.py:473]: Now that I think of it, as I go to implement this, I actually don't even think it's clear what the API should look like.  If you have a column, should you look for n/2 categories, or take two categories and assume the ordering is what is intended for the Wilcoxon test?  In the second case (which would be consistent with the API), there is really no way to distinguish the two models except explicitly by the user saying it so.

[the remark this answers is not in the document]
Comment (ilan-gold) [reply src/multi_condition_comparisions/tl/de.py:473]: I am not familiar with R enough to say what is done there.

[the remark this answers is not in the document]
Comment (grst) [reply src/multi_condition_comparisions/tl/de.py:473]: I'll try to sketch it out in more detail, but I think it'll take until next week

**Comment by grst** [comment]
Closed in favor of #33
